File: 0501-find-mode-in-binary-search-tree/0501-find-mode-in-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public int[] findMode(TreeNode root) {
        List<Integer> inorderr = new ArrayList<>();
        inorder(root, inorderr);

        Map<Integer, Integer> freq = new HashMap<>();
        int maxCount = 0;

        for (int val : inorderr) {
            freq.put(val, freq.getOrDefault(val, 0) + 1);
            maxCount = Math.max(maxCount, freq.get(val));
        }

        List<Integer> result = new ArrayList<>();
        for (Map.Entry<Integer, Integer> entry : freq.entrySet()) {
            if (entry.getValue() == maxCount) {
                result.add(entry.getKey());
            }
        }

        int[] resultArray = new int[result.size()];
        for (int i = 0; i < result.size(); i++) {
            resultArray[i] = result.get(i);
        }

        return resultArray;
    }

    private void inorder(TreeNode root, List<Integer> result) {
        if (root == null) return;
        result.add(root.val);
        inorder(root.left, result);
        inorder(root.right, result);
    }
}

File: 0501-find-mode-in-binary-search-tree/0501-find-mode-in-binary-search-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val === undefined ? 0 : val);
 *     this.left = (left === undefined ? null : left);
 *     this.right = (right === undefined ? null : right);
 * }
 */

/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var findMode = function(root) {
    if (!root) return [];

    // Initialize variables to track mode(s)
    let maxFreq = 0;
    let currentVal = null;
    let currentFreq = 0;

    // Variables to store the result
    let modes = [];

    // Helper function for in-order traversal
    const traverse = (node) => {
        if (!node) return;

        // Visit left subtree
        traverse(node.left);

        // Process current node
        if (node.val === currentVal) {
            currentFreq++;
        } else {
            currentVal = node.val;
            currentFreq = 1;
        }

        // Update max frequency and modes
        if (currentFreq === maxFreq) {
            modes.push(currentVal);
        } else if (currentFreq > maxFreq) {
            maxFreq = currentFreq;
            modes = [currentVal];
        }

        // Visit right subtree
        traverse(node.right);
    };

    // Start in-order traversal
    traverse(root);

    return modes;
};